File: 백트래킹/leetcode1079(Medium)-Letter Tile Possibilities.js
// 백트래킹 : 75ms, 45.3MB
/**
 * @param {string} tiles
 * @return {number}
 */
var numTilePossibilities = function (tiles) {
  const counter = [...tiles].reduce((obj, x) => {
    if (!obj[x]) obj[x] = 0
    obj[x]++
    return obj
  }, {})

  const dfs = (table) => {
    let result = 0
    for (const tile of Object.keys(table)) {
      if (table[tile] <= 0) {
        continue
      }
      table[tile] -= 1
      result += dfs(table) + 1
      table[tile] += 1
    }
    return result
  }

  return dfs(counter)
}

File: 백트래킹/leetcode1079(Medium)-Letter Tile Possibilities.py
# 순열로 완전탐색 : 95ms, 16.8MB
class Solution:
    def numTilePossibilities(self, tiles: str) -> int:
        result = 0
        for i in range(1, len(tiles) + 1):
            permus = list(itertools.permutations(tiles, i))
            pick = set(["".join(list(x)) for x in permus])
            result += len(pick)
        return result

# 백트래킹 : 72ms, 13.9MB (Counter 사용 시 100ms)
class Solution:
    def numTilePossibilities(self, tiles: str) -> int:
        # 백트래킹
        def dfs(table):
            if sum(table.values()) <= 0:
                return 0

            result = 0
            for tile in table.keys():
                if table[tile] <= 0:
                    continue
                table[tile] -= 1
                result += dfs(table) + 1
                table[tile] += 1
            return result

        return dfs(collections.Counter(tiles))